pyspecProcScripts/third_level/protein_QESR.py

```diff
@@ -0,0 +1,156 @@
+    d.set_plot_color(color)
+    fl.next(f"{which_plot} show background subtraction in abs mode")
+    forplot = background.C.integrate(fieldaxis, cumulative=True)
+    forplot.set_plot_color(d.get_plot_color())
+    fl.plot(forplot, alpha=0.2)
+    forplot = d.C.integrate(fieldaxis, cumulative=True)
+    fl.plot(forplot, alpha=0.5, label=label)
+    fl.next(f"{which_plot} absorption, bg. no bl.")
+    # show the same thing on the main plot as a faint line
+    fl.plot(forplot, alpha=0.2, label="%s - no background subtracted"%label)
+    d -= background # subtract background in derivative mode
+    d_abs = d.C.integrate(fieldaxis, cumulative=True)
+    fl.plot(d_abs, alpha=0.5, label=label)
+    # for a protein, we assume we have 1 broad peak that's always over threshold*max
+    right_side = abs(d_abs.data)[-1]
+    # define a peak as anything that rises above threshold*max
+    peaklist = d_abs.contiguous(
+        lambda x: abs(x) > right_side + (abs(x).data.max() - right_side) * threshold
+    )
+    # pull the furthest left and furthest right boundaries of any peaks that we find
+    specrange = (peaklist.ravel().min(), peaklist.ravel().max())
+    fl.next(f"{which_plot} absorption, bg. no bl.")
+    generous_limits = specrange + np.diff(specrange).item() * r_[-pushout, +pushout]
+    for j in r_[np.array(specrange)]:
+        # show where the peaks were identified
+        axvline(x=j, alpha=0.5, color=d.get_plot_color(), ls=":")
+    for j in r_[generous_limits]:
+        # show the actual limits
+        axvline(x=j, alpha=0.5, color=d.get_plot_color(), ls="-")
+    d_baseline = d_abs[
+        fieldaxis,
+        lambda x: np.logical_or(x < generous_limits[0], x > generous_limits[1]),
+    ]
+    fl.next(f"{which_plot} baseline diagnostic")
+    fl.plot(d_baseline, ".", alpha=0.3, label=label, human_units=False)
+    # {{{ we don't have a spline in pyspecdata yet, so just hack it
+    spl = UnivariateSpline(d_baseline.getaxis(fieldaxis), d_baseline.data)
+    polybaseline = d.copy(data=False)
+    polybaseline.data = spl(d.getaxis(fieldaxis))
+    # }}}
+    fl.plot(polybaseline, alpha=0.5, human_units=False)
+    fl.next(f"{which_plot} absorption, bg. no bl.")
+    fl.plot(polybaseline, alpha=0.5)
+    d_abs -= polybaseline
+    d_abs.integrate(fieldaxis, cumulative=True)
+    fl.next("dblint ÷ denom * conversion")
+    d_abs /= QESR_scalefactor(d, calibration_name=calibration_name,
+            diameter_name=diameter_name)
+    final_conc = (
+        d_abs[fieldaxis : (generous_limits[-1], None)].mean(fieldaxis).item()
+    ).real
+    #print("AVerage concentration:", final_conc)
+    d_abs.name("conc").set_units("μM")
+    fl.plot(d_abs, alpha=0.5, label=f"{label}, %0.4f μM" % final_conc)
+    fl.grid()
+    if pickle_file is not None:
+        pickle_vars[label] = final_conc
+        with open(pickle_file, "wb") as fp:
+            pickle_vars = pickle.dump(pickle_vars, fp)
+    return final_conc
```

Comment:
I'm not reading the added scripts in great details, since I figure that you have been using them.

Reply:
I did run the example.